### PR TITLE
Support datetime fields [ch544]

### PR DESCRIPTION
**Pull request by gsvic**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -16,7 +16,7 @@
         <dependency>
             <groupId>io.tiledb</groupId>
             <artifactId>tiledb-java</artifactId>
-            <version>0.3.2</version>
```
Comment (Shelnutt2) [inline pom.xml:19]: I believe this should be put back to 0.3.2?

File: src/main/java/io/prestosql/plugin/tiledb/TileDBMetadata.java [hunk 2]
```diff
@@ -565,11 +564,19 @@ else if (lowerBound < Integer.MIN_VALUE) {
                             }
                             domain.addDimension(new Dimension(localCtx, columnName, classType, new Pair(lowerBound.intValue(), upperBound.intValue()), extent.intValue()));
                             break;
+                        case TILEDB_DATETIME_MS:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBMetadata.java:567]: You are missing the items below MS?

File: src/main/java/io/prestosql/plugin/tiledb/TileDBModule.java [hunk 3]
```diff
@@ -125,6 +128,16 @@ public static Type prestoTypeFromTileDBType(Datatype type) throws TileDBError
                 return REAL;
             case TILEDB_FLOAT64:
                 return DOUBLE;
+            case TILEDB_DATETIME_MS:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBModule.java:131]: Add the types below `MS` (AS/NS/US etc)

File: src/main/java/io/prestosql/plugin/tiledb/TileDBModule.java [hunk 4]
```diff
@@ -204,6 +220,8 @@ else if (type.equals(DATE)) {
                 }
                 return new ArrayList<Integer>();
             }
+            case TILEDB_DATETIME_MS:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBModule.java:223]: Do you need to support other DATETIMES here? Where is `getJavaListForType` used?

File: pom.xml (hunk 1 above)
Comment (gsvic) [reply pom.xml]: Correct, I reverted that back to 0.3.2.

File: src/main/java/io/prestosql/plugin/tiledb/TileDBPageSink.java [hunk 5]
```diff
@@ -386,8 +392,11 @@ else if (isVarcharType(type) || isCharType(type)) {
         else if (VARBINARY.equals(type)) {
             columnBuffer.setItem(bufferPosition, type.getSlice(block, position).getBytes());
         }
-        else if (DATE.equals(type)) { // NOTE: this is not used because we make all date columns a varchar type
-            columnBuffer.setItem(bufferPosition, DATE_FORMATTER.print(TimeUnit.DAYS.toMillis(type.getLong(block, position))));
+        else if (DATE.equals(type)) {
+            columnBuffer.setItem(bufferPosition, type.getLong(block, position));
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBPageSink.java:396]: You can't just set the date/timestamp directly. You have to do conversions to the real tiledb datatype. I.e. if the field is `TILEDB_DATETIME_YEAR` and you set it to a date it'll be stored wrong. I.e. if the user inserts `2020-01-1`, the expected value to be stored is only `50`.

File: src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java [hunk 6]
```diff
@@ -544,6 +545,8 @@ private static Object tiledbValueMin(Object a, Object b, Datatype type) throws T
                 return (short) a < (short) b ? a : b;
             case TILEDB_INT32:
                 return (int) a < (int) b ? a : b;
+            case TILEDB_DATETIME_MS:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java:548]: Need to support all datetimes

File: src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java [hunk 7]
```diff
@@ -577,6 +580,8 @@ private static Object tiledbValueMax(Object a, Object b, Datatype type) throws T
                 return (short) a > (short) b ? a : b;
             case TILEDB_INT32:
                 return (int) a > (int) b ? a : b;
+            case TILEDB_DATETIME_MS:
+            case TILEDB_DATETIME_DAY:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java:584]: Need to support all datetimes

File: src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java [hunk 8]
```diff
@@ -607,6 +612,8 @@ private static Object addEpsilon(Object value, Datatype type) throws TileDBError
                 return ((short) value) < Short.MAX_VALUE ? ((short) value + 1) : value;
             case TILEDB_INT32:
                 return ((int) value) < Integer.MAX_VALUE ? ((int) value + 1) : value;
+            case TILEDB_DATETIME_MS:
+            case TILEDB_DATETIME_DAY:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java:616]: Need to support all datetimes

File: src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java [hunk 9]
```diff
@@ -647,6 +654,8 @@ private static Object subtractEpsilon(Object value, Datatype type) throws TileDB
                 return ((int) value) > ((int) Short.MIN_VALUE - 1) ? ((int) value - 1) : value;
             case TILEDB_UINT32:
                 return ((long) value) > ((long) Integer.MIN_VALUE - 1) ? ((long) value - 1) : value;
+            case TILEDB_DATETIME_MS:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java:657]: Need to support all datetimes

File: src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java [hunk 10]
```diff
@@ -953,6 +962,32 @@ public long getLong(int field)
...
+            case TILEDB_DATETIME_MONTH: {
+                value = Timestamp.valueOf(new Timestamp(0).toLocalDateTime().plusMonths(((long[]) fieldArray)[index])).getTime();
+                break;
+            }
+            case TILEDB_DATETIME_YEAR: {
+                value = Timestamp.valueOf(new Timestamp(0).toLocalDateTime().plusYears(((long[]) fieldArray)[index])).getTime();
+                break;
+            }
+            case TILEDB_DATETIME_MS:
+            case TILEDB_DATETIME_DAY:
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java:990]: Need to support all datetimes, missing `AS`/`NS/`US/`FS`, etc.

File: src/main/java/io/prestosql/plugin/tiledb/TileDBModule.java (hunk 4 above)
Comment (gsvic) [reply src/main/java/io/prestosql/plugin/tiledb/TileDBModule.java:223]: It looks like that this is not used at the moment, there is no caller function in the codebase

File: src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java [hunk 11]
```diff
@@ -974,47 +1008,47 @@ public long getLong(int field)
...
+                value = new Timestamp(0).toInstant().atOffset(ZoneOffset.UTC).plusNanos((long) (((long[]) fieldArray)[index] * 0.001)).toInstant().toEpochMilli();
                 break;
             }
             case TILEDB_DATETIME_PS: {
-                value = Timestamp.valueOf((new Timestamp(0).toLocalDateTime().plusNanos((long) (((long[]) fieldArray)[index] * 0.01)))).getTime();
+                value = new Timestamp(0).toInstant().atOffset(ZoneOffset.UTC).plusNanos((long) (((long[]) fieldArray)[index] * 0.01)).toInstant().toEpochMilli();
                 break;
             }
             case TILEDB_DATETIME_NS: {
                 value = Timestamp.valueOf(new Timestamp(0).toLocalDateTime().plusNanos(((long[]) fieldArray)[index])).getTime();
```
Comment (Shelnutt2) [inline src/main/java/io/prestosql/plugin/tiledb/TileDBRecordCursor.java:1023]: Looks like this was missed for the UTC change?

**Review by Shelnutt2 — COMMENTED** [review]
two minor adjustments, then I think there is a few `println` that need to be removed tomorrow, after that it should be good to go!

**Review by Shelnutt2 — APPROVED** [review]

**Pull request merged**